core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -442,60 +444,39 @@ sealed abstract class Doc extends Product with Serializable {
     loop(Left(this) :: Nil, Doc.empty)
   }
 
-  /**
-   * This method is similar to flatten, but returns None if no
-   * flattening was needed (i.e. if no newlines or unions were present).
-   *
-   * As with flatten, the resulting Doc (if any) will never render any
-   * newlines, no matter what width is used.
-   */
-  def flattenOption: Option[Doc] = {
-    val res = flattenBoolean
-    if (res._2) Some(res._1) else None
-  }
+  private[paiges] def tryFlatten: Doc = flatten getOrElse this
 
-  /**
-   * Convert this Doc to a single-line representation.
-   *
-   * All newlines are replaced with spaces (and optional indentation
-   * is ignored). The resulting Doc will never render any newlines, no
-   * matter what width is used.
-   */
-  def flatten: Doc = flattenOption getOrElse this
-
-  // return the flattened doc, and if it is different
-  private def flattenBoolean: (Doc, Boolean) = {
-
-    type DB = (Doc, Boolean)
+  // Try to put the document on a single line.  Will succeed iff there are no
+  // hard line breaks.
+  // Invariant: If flatten(d) == Some(d') then d' has no FlatAlt nodes
+  private[paiges] def flatten: Option[Doc] = {
+    sealed trait Flatten extends Product with Serializable
```

Comment:
this change seems orthogonal to the rest. Can we pull out a small change that just introduces this `Flatten` type to clean the code up?

Reply:
To clarify, you'd like this change without the new `Line` constructor?  Then the `No` branch is unused, so I didn't think of this as totally orthogonal.